src/vs/workbench/parts/preferences/common/keybindingsEditorModel.ts

```diff
@@ -102,33 +102,66 @@ export class KeybindingsEditorModel extends EditorModel {
 		return this.fetchKeybindingItems(sortByPrecedence ? this._keybindingItemsSortedByPrecedence : this._keybindingItems, searchValue, quoteAtFirstChar && quoteAtLastChar);
 	}
 
+	isSourceFilterApplied(words: string[]): boolean {
+		return (words.length) && (words[0].trim().toLowerCase() === '@source:');
+	}
+
+	getSourceFilterValue(words: string[]): string {
+		return words[1] || '';
+	}
+
 	private fetchKeybindingItems(keybindingItems: IKeybindingItem[], searchValue: string, completeMatch: boolean): IKeybindingItemEntry[] {
 		if (!searchValue) {
 			return keybindingItems.map(keybindingItem => ({ id: KeybindingsEditorModel.getId(keybindingItem), keybindingItem, templateId: KEYBINDING_ENTRY_TEMPLATE_ID }));
 		}
 
 		const result: IKeybindingItemEntry[] = [];
 		const words = searchValue.split(' ');
+
+		const sourceFilterApplied: boolean = this.isSourceFilterApplied(words);
+		let sourceFilterSearchValue: string;
+		if (sourceFilterApplied) {
+			sourceFilterSearchValue = this.getSourceFilterValue(words);
+		}
+
 		const keybindingWords = this.splitKeybindingWords(words);
 		for (const keybindingItem of keybindingItems) {
-			let keybindingMatches = new KeybindingItemMatches(this.modifierLabels, keybindingItem, searchValue, words, keybindingWords, completeMatch);
-			if (keybindingMatches.commandIdMatches
-				|| keybindingMatches.commandLabelMatches
-				|| keybindingMatches.commandDefaultLabelMatches
-				|| keybindingMatches.sourceMatches
-				|| keybindingMatches.whenMatches
-				|| keybindingMatches.keybindingMatches) {
-				result.push({
-					id: KeybindingsEditorModel.getId(keybindingItem),
-					templateId: KEYBINDING_ENTRY_TEMPLATE_ID,
-					commandLabelMatches: keybindingMatches.commandLabelMatches,
-					commandDefaultLabelMatches: keybindingMatches.commandDefaultLabelMatches,
-					keybindingItem,
-					keybindingMatches: keybindingMatches.keybindingMatches,
-					commandIdMatches: keybindingMatches.commandIdMatches,
-					sourceMatches: keybindingMatches.sourceMatches,
-					whenMatches: keybindingMatches.whenMatches
-				});
+			if (sourceFilterApplied) {
```

Comment:
1. Is there a better way to push results based on whether `@source` filter is applied or not?
2. Can we move refactor it into on method which checks for respective matches and another one which creates the final `result` array?

Reply:
@shobhitchittora You can have two methods one for filtering based on source and the other does full text search and call either of them by checking the search text. It can look something like follows

```typescript
if (/@source:/i.test(searchValue)) {
return this.filterBySource(searchValue);
}
return this.filterByText(searchValue);
```